numpy/_core/src/multiarray/unique.cpp

```diff
@@ -0,0 +1,189 @@
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+
+#include <Python.h>
+
+#include <unordered_set>
+#include <functional>
+
+#include "numpy/arrayobject.h"
+#include "numpy/npy_common.h"
+
+
+// This is to use RAII pattern to handle cpp exceptions while avoiding memory leaks.
+// Adapted from https://stackoverflow.com/a/25510879/2536294
+template <typename F>
+struct FinalAction {
+    FinalAction(F f) : clean_{f} {}
+    ~FinalAction() { clean_(); }
+  private:
+    F clean_;
+};
+
+template <typename F>
+FinalAction<F> finally(F f) {
+    return FinalAction<F>(f);
+}
+
+template<typename T>
+PyObject* unique(PyArrayObject *self)
+{
+    /* This function takes a numpy array and returns a numpy array containing
+    the unique values.
+
+    It assumes the numpy array includes data that can be viewed as unsigned integers
+    of a certain size (sizeof(T)).
+
+    It doesn't need to know the actual type, since it needs to find unique values
+    among binary representations of the input data. This means it won't apply to
+    custom or complicated dtypes or string values.
+    */
+    NpyIter* iter;
+    NpyIter_IterNextFunc *iternext;
+    char** dataptr;
+    npy_intp* strideptr,* innersizeptr;
+    std::unordered_set<T> hashset;
+
+    iter = NpyIter_New(self, NPY_ITER_READONLY|
+                             NPY_ITER_EXTERNAL_LOOP|
+                             NPY_ITER_REFS_OK,
+                        NPY_KEEPORDER, NPY_NO_CASTING,
+                        NULL);
+    // Making sure the iterator is deallocated when the function returns, with
+    // or w/o an exception
+    auto iter_dealloc = finally([&]() { NpyIter_Deallocate(iter); });
+    if (iter == NULL) {
+        return Py_None;
+    }
+
+    iternext = NpyIter_GetIterNext(iter, NULL);
+    if (iternext == NULL) {
+        return Py_None;
+    }
+    dataptr = NpyIter_GetDataPtrArray(iter);
+    strideptr = NpyIter_GetInnerStrideArray(iter);
+    innersizeptr = NpyIter_GetInnerLoopSizePtr(iter);
+
+    // first we put the data in a hash map
+    do {
+        char* data = *dataptr;
+        npy_intp stride = *strideptr;
+        npy_intp count = *innersizeptr;
+
+        while (count--) {
+            hashset.insert(*((T *) data));
+            data += stride;
+        }
+    } while(iternext(iter));
+
+    // then we iterate through the map's keys to get the unique values
+    T* res = new T[hashset.size()];
```

Comment:
does this need any refcounting?

Reply:
What we need is that we have to make sure two things:
1. We need to ensure we clean up `iter` (call `NpyIter_Dealloc`) even if there are C++ exceptions.  `new` can raise such an exception (but I am sure many things here can).
2. If a C++ exception occurs, we need to convert it into some reasonable Python error.  In most cases these should be critical errors (i.e. `RuntimeError` is OK I think), although distinguishing `MemoryError` may be good.
   We _might_ have a pattern for this in the fft ufunc code.

Otherwise, I think I can live with this.  We could move this `unique` function onto the dtype.
A question for that would be whether we can refactor the function so that it actually becomes:
```
dtype.get_unique(&update_function, &finalize_function, &auxdata)

do {
     if (update_function(context, data, stride, count, auxdata) < 0) {
          break;  /* ouch, error */
     }
} while (iternext)

void *res = finalize_function(context, auxdata);
```
I _think_ that is not a problem (and I suspect required to disect out knowing about NumPy arrays and creating a stable C-API).

TL;DR:  Should we meet up and walk through the code once (if just to force me to  look at it closer?).  I think I am fine with this, although I might want to at least put a hint to what I wrote above into the code to tell people how we may achieve that users can implement unique for their dtypes.

(I think I might like to already split it into that 1-D `update_func` and a `finalize_func`, just because that is such a typical pattern in NumPy to organize things around 1-D loops.)